### PR TITLE
add app checkout provider

### PR DESCRIPTION
**Pull request by jfeaver**
Potential issue here: 0 tests added to effective_orders :(

@matt-riemer: I'm submitting this PR now in case you'd like to get this feature into the EffectiveOrders 2.0 release. If it's okay to hold back, I can write tests and add them to the PR.

### CODE REVIEW

File: lib/effective_orders/app_checkout_service.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module EffectiveOrders
+  class AppCheckoutService
+    def self.call(order:)
```
Comment (matt-riemer) [inline lib/effective_orders/app_checkout_service.rb]: colon here?

File: lib/effective_orders/app_checkout_service.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      new(order).tap(&:call)
+    end
+
+    attr_reader :order
+
+    def initialize(order)
+      @order = order
+    end
+
+    def call
```
Comment (matt-riemer) [inline lib/effective_orders/app_checkout_service.rb]: should this check for NotImplementedError as well?

File: app/controllers/effective/providers/app_checkout.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+          order_purchased(payment_details(checkout))
+        else
+          flash = EffectiveOrders.app_checkout[:declined_flash]
+          order_declined(payment_details(checkout), nil, flash: flash)
+        end
+      end
+
+      def payment_details(checkout)
+        default = 'App Checkout'
+        if checkout.respond_to?(:payment_details)
```
Comment (matt-riemer) [inline app/controllers/effective/providers/app_checkout.rb]: the service is always going to respond to payment_details, as its base class does.

File: lib/effective_orders/engine.rb [hunk 4]
```diff
@@ -90,6 +90,17 @@ class Engine < ::Rails::Engine
...
+    initializer 'effective_orders.app_checkout_config_validation', :after => :load_config_initializers do
+      if EffectiveOrders.app_checkout_enabled
+        unless EffectiveOrders.app_checkout.is_a?(Hash)
+          raise ArgumentError, "expected EffectiveOrders.app_checkout to be a Hash but it is a #{EffectiveOrders.app_checkout.class}"
+        end
+        missing = EffectiveOrders.app_checkout.select {|_config, value| value.blank? }
+
+        raise "Missing effective_orders App Checkout configuration values: #{missing.keys.join(', ')}" if missing.present?
+      end
+    end
```
Comment (matt-riemer) [inline lib/effective_orders/engine.rb]: Consider adding a check for if app_checkout[:service].present? make sure it's a kind_of?(EffectiveOrders::AppCheckoutService}

File: lib/effective_orders/app_checkout_service.rb (hunk 1 above)
Comment (jfeaver) [inline lib/effective_orders/app_checkout_service.rb]: Oops, I should code this differently for Ruby backward compatibility but the colon is intentional: https://robots.thoughtbot.com/ruby-2-keyword-arguments#required-keyword-arguments

File: lib/effective_orders/app_checkout_service.rb (hunk 2 above)
Comment (jfeaver) [inline lib/effective_orders/app_checkout_service.rb]: The call method isn't strictly required but is likely to be included. A payment could be successful by default and not require payment (perhaps). I could add a `NotImplementedError` and just allow the user to implement an empty method if they really want.

File: app/controllers/effective/providers/app_checkout.rb (hunk 3 above)
Comment (jfeaver) [inline app/controllers/effective/providers/app_checkout.rb]: Remember that the service object does not need to inherit from the base class that I provide. The `payment_details` method should probably be included in the README.

File: lib/effective_orders/engine.rb (hunk 4 above)
Comment (jfeaver) [inline lib/effective_orders/engine.rb]: It doesn't need to pass that check. It does need to `respond_to?(:call)` though.

**Pull request merged**